# Geoprocessing NetCDF Files

## Import libraries

In [1]:
import arcpy
from arcpy.ia import*

In [2]:
# connect to GIS
from arcgis.gis import GIS

## Login ArcGIS online

In [3]:
# Prompt user to provide username and password
import getpass
username = input('Enter username: ')
password = getpass.getpass("Enter your password: ") 
gis = GIS("https://udistritalfjc.maps.arcgis.com/home", username, password)

Enter username: camendezv_UDFJC
Enter your password: ········


## Import the netCDF files with precipitation data

In [4]:
# Precipitation

# Model CCCma-CanESM2
mdRaster_CanESM2_pr = Raster(r"C:\Users\57320\Dropbox\Mi PC (LAPTOP-5K2HI422)\Documents\ArcGIS for Climate Change\precipitation\pr_SAM-20_CCCma-CanESM2_rcp45_r1i1p1_INPE-Eta_v1_mon_202101-203012.nc", True)
# Model MIROC-MIROC5
mdRaster_MIROC5_pr = Raster(r"C:\Users\57320\Dropbox\Mi PC (LAPTOP-5K2HI422)\Documents\ArcGIS for Climate Change\precipitation\pr_SAM-20_MIROC-MIROC5_rcp45_r1i1p1_INPE-Eta_v1_mon_202101-203012.nc", True)
# Model MOHC-HadGEM2-ES
mdRaster_HadGEM2_pr = Raster(r"C:\Users\57320\Dropbox\Mi PC (LAPTOP-5K2HI422)\Documents\ArcGIS for Climate Change\precipitation\pr_SAM-20_MOHC-HadGEM2-ES_rcp45_r1i1p1_INPE-Eta_v1_mon_202101-203012.nc", True)

# Near Surface Air Temperature

# Model CCCma-CanESM2
mdRaster_CanESM2_tas = Raster(r"C:\Users\57320\Dropbox\Mi PC (LAPTOP-5K2HI422)\Documents\ArcGIS for Climate Change\temperature\tas_SAM-20_CCCma-CanESM2_rcp45_r1i1p1_INPE-Eta_v1_mon_202101-203012.nc", True)
# Model MIROC-MIROC5
mdRaster_MIROC5_tas = Raster(r"C:\Users\57320\Dropbox\Mi PC (LAPTOP-5K2HI422)\Documents\ArcGIS for Climate Change\temperature\tas_SAM-20_MIROC-MIROC5_rcp45_r1i1p1_INPE-Eta_v1_mon_202101-203012.nc", True)
# Model MOHC-HadGEM2-ES
mdRaster_HadGEM2_tas = Raster(r"C:\Users\57320\Dropbox\Mi PC (LAPTOP-5K2HI422)\Documents\ArcGIS for Climate Change\temperature\tas_SAM-20_MOHC-HadGEM2-ES_rcp45_r1i1p1_INPE-Eta_v1_mon_202101-203012.nc", True)

## Check the properties and verify  if the data is multidimensional

### Precipitation data

#### Model CCCma-CanESM2

In [5]:
mdRaster_CanESM2_pr.isMultidimensional

True

In [6]:
mdRaster_CanESM2_pr.mdinfo

'{"variables":[{"name":"pr","unit":"kg m-2 s-1","attributes":{"standard_name":"precipitation_flux","long_name":"Precipitation","cell_methods":"time: mean","_FillValue":1.0000000200408773e+20},"dimensions":[{"name":"StdTime","field":"StdTime","unit":"ISO8601","interval":29.5,"hasRegularIntervals":false,"intervalUnit":"Days","extent":["2021-01-01T00:00:00","2031-01-01T00:00:00"],"hasRanges":true,"values":[["2021-01-01T00:00:00","2021-02-01T00:00:00"],["2021-02-01T00:00:00","2021-03-01T00:00:00"],["2021-03-01T00:00:00","2021-04-01T00:00:00"],["2021-04-01T00:00:00","2021-05-01T00:00:00"],["2021-05-01T00:00:00","2021-06-01T00:00:00"],["2021-06-01T00:00:00","2021-07-01T00:00:00"],["2021-07-01T00:00:00","2021-08-01T00:00:00"],["2021-08-01T00:00:00","2021-09-01T00:00:00"],["2021-09-01T00:00:00","2021-10-01T00:00:00"],["2021-10-01T00:00:00","2021-11-01T00:00:00"],["2021-11-01T00:00:00","2021-12-01T00:00:00"],["2021-12-01T00:00:00","2022-01-01T00:00:00"],["2022-01-01T00:00:00","2022-02-01T00:00:

In [7]:
mdRaster_CanESM2_pr.extent

<Extent object at 0x29585bb1d30[0x29585bb1c60]>

In [8]:
print(mdRaster_CanESM2_pr.variables)
print(mdRaster_CanESM2_pr.variableNames)
print(mdRaster_CanESM2_pr.getDimensionNames('pr'))

['pr(StdTime=120)']
['pr']
['StdTime']


In [10]:
#Consult the statistics
mdRaster_CanESM2_pr.getStatistics('pr')

[{'min': 0.0, 'max': 0.0007541565573774278, 'mean': 2.9414161154831338e-05, 'standardDeviation': 3.655579805296284e-05, 'sum': 0.0, 'skipX': 1, 'skipY': 1, 'count': 0.0}]

#### Model MIROC-MIROC5

In [40]:
mdRaster_MIROC5_pr.isMultidimensional

True

In [41]:
mdRaster_MIROC5_pr.mdinfo

'{"variables":[{"name":"pr","unit":"kg m-2 s-1","attributes":{"standard_name":"precipitation_flux","long_name":"Precipitation","cell_methods":"time: mean","_FillValue":1.0000000200408773e+20},"dimensions":[{"name":"StdTime","field":"StdTime","unit":"ISO8601","interval":29.5,"hasRegularIntervals":false,"intervalUnit":"Days","extent":["2021-01-01T00:00:00","2031-01-01T00:00:00"],"hasRanges":true,"values":[["2021-01-01T00:00:00","2021-02-01T00:00:00"],["2021-02-01T00:00:00","2021-03-01T00:00:00"],["2021-03-01T00:00:00","2021-04-01T00:00:00"],["2021-04-01T00:00:00","2021-05-01T00:00:00"],["2021-05-01T00:00:00","2021-06-01T00:00:00"],["2021-06-01T00:00:00","2021-07-01T00:00:00"],["2021-07-01T00:00:00","2021-08-01T00:00:00"],["2021-08-01T00:00:00","2021-09-01T00:00:00"],["2021-09-01T00:00:00","2021-10-01T00:00:00"],["2021-10-01T00:00:00","2021-11-01T00:00:00"],["2021-11-01T00:00:00","2021-12-01T00:00:00"],["2021-12-01T00:00:00","2022-01-01T00:00:00"],["2022-01-01T00:00:00","2022-02-01T00:00:

In [42]:
mdRaster_MIROC5_pr.extent

<Extent object at 0x29588c1f370[0x29588c1f2a0]>

In [43]:
print(mdRaster_MIROC5_pr.variables)
print(mdRaster_MIROC5_pr.variableNames)
print(mdRaster_MIROC5_pr.getDimensionNames('pr'))

['pr(StdTime=120)']
['pr']
['StdTime']


In [44]:
#Consult the statistics
mdRaster_MIROC5_pr.getStatistics('pr')

[{'min': 0.0, 'max': 0.0007419097819365561, 'mean': 4.1113089014453676e-05, 'standardDeviation': 4.9780031471888254e-05, 'sum': 0.0, 'skipX': 1, 'skipY': 1, 'count': 0.0}]

####  Model MOHC-HadGEM2-ES

In [45]:
mdRaster_HadGEM2_pr.isMultidimensional

True

In [46]:
mdRaster_HadGEM2_pr.mdinfo

'{"variables":[{"name":"pr","unit":"kg m-2 s-1","attributes":{"cell_methods":"time: mean","long_name":"Precipitation","standard_name":"precipitation_flux","_FillValue":1.0000000200408773e+20},"dimensions":[{"name":"StdTime","field":"StdTime","unit":"ISO8601","interval":29.5,"hasRegularIntervals":false,"intervalUnit":"Days","extent":["2021-01-01T00:00:00","2031-01-01T00:00:00"],"hasRanges":true,"values":[["2021-01-01T00:00:00","2021-02-01T00:00:00"],["2021-02-01T00:00:00","2021-03-01T00:00:00"],["2021-03-01T00:00:00","2021-04-01T00:00:00"],["2021-04-01T00:00:00","2021-05-01T00:00:00"],["2021-05-01T00:00:00","2021-06-01T00:00:00"],["2021-06-01T00:00:00","2021-07-01T00:00:00"],["2021-07-01T00:00:00","2021-08-01T00:00:00"],["2021-08-01T00:00:00","2021-09-01T00:00:00"],["2021-09-01T00:00:00","2021-10-01T00:00:00"],["2021-10-01T00:00:00","2021-11-01T00:00:00"],["2021-11-01T00:00:00","2021-12-01T00:00:00"],["2021-12-01T00:00:00","2022-01-01T00:00:00"],["2022-01-01T00:00:00","2022-02-01T00:00:

In [47]:
mdRaster_HadGEM2_pr.extent

<Extent object at 0x29585bca790[0x29585bca5d0]>

In [48]:
print(mdRaster_HadGEM2_pr.variables)
print(mdRaster_HadGEM2_pr.variableNames)
print(mdRaster_HadGEM2_pr.getDimensionNames('pr'))

['pr(StdTime=120)']
['pr']
['StdTime']


In [49]:
#Consult the statistics
mdRaster_HadGEM2_pr.getStatistics('pr')

[{'min': 0.0, 'max': 0.0011640851153060794, 'mean': 2.7521296154294152e-05, 'standardDeviation': 3.269226990630866e-05, 'sum': 0.0, 'skipX': 1, 'skipY': 1, 'count': 0.0}]

### Near Surface Air Temperature

#### Model CCCma-CanESM2

In [50]:
mdRaster_CanESM2_tas.isMultidimensional

True

In [51]:
mdRaster_CanESM2_tas.mdinfo

'{"variables":[{"name":"tas","unit":"K","attributes":{"standard_name":"air_temperature","long_name":"Near-Surface Air Temperature","cell_methods":"time: mean","_FillValue":1.0000000200408773e+20},"dimensions":[{"name":"StdTime","field":"StdTime","unit":"ISO8601","interval":29.5,"hasRegularIntervals":false,"intervalUnit":"Days","extent":["2021-01-01T00:00:00","2031-01-01T00:00:00"],"hasRanges":true,"values":[["2021-01-01T00:00:00","2021-02-01T00:00:00"],["2021-02-01T00:00:00","2021-03-01T00:00:00"],["2021-03-01T00:00:00","2021-04-01T00:00:00"],["2021-04-01T00:00:00","2021-05-01T00:00:00"],["2021-05-01T00:00:00","2021-06-01T00:00:00"],["2021-06-01T00:00:00","2021-07-01T00:00:00"],["2021-07-01T00:00:00","2021-08-01T00:00:00"],["2021-08-01T00:00:00","2021-09-01T00:00:00"],["2021-09-01T00:00:00","2021-10-01T00:00:00"],["2021-10-01T00:00:00","2021-11-01T00:00:00"],["2021-11-01T00:00:00","2021-12-01T00:00:00"],["2021-12-01T00:00:00","2022-01-01T00:00:00"],["2022-01-01T00:00:00","2022-02-01T00

In [52]:
mdRaster_CanESM2_tas.extent

<Extent object at 0x29588c1f340[0x29588c1f360]>

In [53]:
print(mdRaster_CanESM2_tas.variables)
print(mdRaster_CanESM2_tas.variableNames)
print(mdRaster_CanESM2_tas.getDimensionNames('tas'))

['tas(StdTime=120)']
['tas']
['StdTime']


In [54]:
#Consult the statistics
mdRaster_CanESM2_tas.getStatistics('tas')

[{'min': 265.037109375, 'max': 307.15289306640625, 'mean': 295.0058663573418, 'standardDeviation': 6.022663791452304, 'sum': 0.0, 'skipX': 1, 'skipY': 1, 'count': 0.0}]

#### Model MIROC-MIROC5

In [55]:
mdRaster_MIROC5_tas.isMultidimensional

True

In [56]:
mdRaster_MIROC5_tas.mdinfo

'{"variables":[{"name":"tas","unit":"K","attributes":{"standard_name":"air_temperature","long_name":"Near-Surface Air Temperature","cell_methods":"time: mean","_FillValue":1.0000000200408773e+20},"dimensions":[{"name":"StdTime","field":"StdTime","unit":"ISO8601","interval":29.5,"hasRegularIntervals":false,"intervalUnit":"Days","extent":["2021-01-01T00:00:00","2031-01-01T00:00:00"],"hasRanges":true,"values":[["2021-01-01T00:00:00","2021-02-01T00:00:00"],["2021-02-01T00:00:00","2021-03-01T00:00:00"],["2021-03-01T00:00:00","2021-04-01T00:00:00"],["2021-04-01T00:00:00","2021-05-01T00:00:00"],["2021-05-01T00:00:00","2021-06-01T00:00:00"],["2021-06-01T00:00:00","2021-07-01T00:00:00"],["2021-07-01T00:00:00","2021-08-01T00:00:00"],["2021-08-01T00:00:00","2021-09-01T00:00:00"],["2021-09-01T00:00:00","2021-10-01T00:00:00"],["2021-10-01T00:00:00","2021-11-01T00:00:00"],["2021-11-01T00:00:00","2021-12-01T00:00:00"],["2021-12-01T00:00:00","2022-01-01T00:00:00"],["2022-01-01T00:00:00","2022-02-01T00

In [57]:
mdRaster_MIROC5_tas.extent

<Extent object at 0x29585bca0a0[0x29585bca780]>

In [58]:
print(mdRaster_MIROC5_tas.variables)
print(mdRaster_MIROC5_tas.variableNames)
print(mdRaster_MIROC5_tas.getDimensionNames('tas'))

['tas(StdTime=120)']
['tas']
['StdTime']


In [59]:
#Consult the statistics
mdRaster_MIROC5_tas.getStatistics('tas')

[{'min': 258.97509765625, 'max': 303.2293701171875, 'mean': 294.66919364887895, 'standardDeviation': 5.278882579542881, 'sum': 0.0, 'skipX': 1, 'skipY': 1, 'count': 0.0}]

####  Model MOHC-HadGEM2-ES

In [35]:
mdRaster_HadGEM2_tas.isMultidimensional

True

In [60]:
mdRaster_HadGEM2_tas.mdinfo

'{"variables":[{"name":"tas","unit":"K","attributes":{"cell_methods":"time: mean","long_name":"Near-Surface Air Temperature","standard_name":"air_temperature","_FillValue":1.0000000200408773e+20},"dimensions":[{"name":"StdTime","field":"StdTime","unit":"ISO8601","interval":29.5,"hasRegularIntervals":false,"intervalUnit":"Days","extent":["2021-01-01T00:00:00","2031-01-01T00:00:00"],"hasRanges":true,"values":[["2021-01-01T00:00:00","2021-02-01T00:00:00"],["2021-02-01T00:00:00","2021-03-01T00:00:00"],["2021-03-01T00:00:00","2021-04-01T00:00:00"],["2021-04-01T00:00:00","2021-05-01T00:00:00"],["2021-05-01T00:00:00","2021-06-01T00:00:00"],["2021-06-01T00:00:00","2021-07-01T00:00:00"],["2021-07-01T00:00:00","2021-08-01T00:00:00"],["2021-08-01T00:00:00","2021-09-01T00:00:00"],["2021-09-01T00:00:00","2021-10-01T00:00:00"],["2021-10-01T00:00:00","2021-11-01T00:00:00"],["2021-11-01T00:00:00","2021-12-01T00:00:00"],["2021-12-01T00:00:00","2022-01-01T00:00:00"],["2022-01-01T00:00:00","2022-02-01T00

In [61]:
mdRaster_HadGEM2_tas.extent

<Extent object at 0x29588c1f2b0[0x29588c1f2d0]>

In [62]:
print(mdRaster_HadGEM2_tas.variables)
print(mdRaster_HadGEM2_tas.variableNames)
print(mdRaster_HadGEM2_tas.getDimensionNames('tas'))

['tas(StdTime=120)']
['tas']
['StdTime']


In [63]:
#Consult the statistics
mdRaster_HadGEM2_tas.getStatistics('tas')

[{'min': 264.39520263671875, 'max': 306.792724609375, 'mean': 295.10375581101215, 'standardDeviation': 4.959168387540001, 'sum': 0.0, 'skipX': 1, 'skipY': 1, 'count': 0.0}]

## Conversion of units

### Convert units `kg*m-2*s-1` to `mm/month` in precipitation data

The default values of precipitation in NetCDF files are in `kg*m-2*s-1`, for this reason it´s necessary to convert this values to `mm/day` and then in `mm/month`

the constant 86400, convert `kg*m-2*s-1` in `mm/day`. However, the interval slices are 29.5 days. Finally, multiply 86400*29.5 = 2548800, it´s the final constant 

In [65]:
# Model CCCma-CanESM2
with arcpy.EnvManager(scratchWorkspace = r"C:\Users\57320\Dropbox\Mi PC (LAPTOP-5K2HI422)\Documents\ArcGIS for Climate Change\precipitation"):
    out_raster = arcpy.sa.Times(in_raster_or_constant1 = mdRaster_CanESM2_pr, in_raster_or_constant2 = 2548800)
    out_raster.save(r"C:\Users\57320\Dropbox\Mi PC (LAPTOP-5K2HI422)\Documents\ArcGIS for Climate Change\precipitation\CCCma-CanESM2_prcp.crf")

In [ ]:
# Model MIROC-MIROC5
with arcpy.EnvManager(scratchWorkspace = r"C:\Users\57320\Dropbox\Mi PC (LAPTOP-5K2HI422)\Documents\ArcGIS for Climate Change\precipitation"):
    out_raster = arcpy.sa.Times(in_raster_or_constant1 = mdRaster_MIROC5_pr, in_raster_or_constant2 = 2548800)
    out_raster.save(r"C:\Users\57320\Dropbox\Mi PC (LAPTOP-5K2HI422)\Documents\ArcGIS for Climate Change\precipitation\MIROC-MIROC5_prcp.crf")

In [66]:
# Model MOHC-HadGEM2-ES
with arcpy.EnvManager(scratchWorkspace = r"C:\Users\57320\Dropbox\Mi PC (LAPTOP-5K2HI422)\Documents\ArcGIS for Climate Change\precipitation"):
    out_raster = arcpy.sa.Times(in_raster_or_constant1 = mdRaster_HadGEM2_pr, in_raster_or_constant2 = 2548800)
    out_raster.save(r"C:\Users\57320\Dropbox\Mi PC (LAPTOP-5K2HI422)\Documents\ArcGIS for Climate Change\precipitation\MOHC-HadGEM2-ES_prcp.crf")

### Convert `Kelvin` to `Celsius` units in Near Surface Air Temperature Data

The default values of Near Surface Air Temperature in NetCDF files are in `Kelvin`, for this reason it´s necessary to convert this values to `Celsius`. The raster function `Unit Conversion` allow to convert Kelvin in Celsius 


In [76]:
# Model CCCma-CanESM2
from arcpy.ia import *
out_unit_raster = UnitConversion(mdRaster_CanESM2_tas,"Kelvin", "Celsius")
out_unit_raster.save(r"C:\Users\57320\Dropbox\Mi PC (LAPTOP-5K2HI422)\Documents\ArcGIS for Climate Change\temperature\CCCma-CanESM2_tas.crf")

In [77]:
# Model MIROC-MIROC5
from arcpy.ia import *
out_unit_raster = UnitConversion(mdRaster_MIROC5_tas,"Kelvin", "Celsius")
out_unit_raster.save(r"C:\Users\57320\Dropbox\Mi PC (LAPTOP-5K2HI422)\Documents\ArcGIS for Climate Change\temperature\MIROC-MIROC5_tas.crf")

In [78]:
# Model MOHC-HadGEM2-ES
from arcpy.ia import *
out_unit_raster = UnitConversion(mdRaster_HadGEM2_tas,"Kelvin", "Celsius")
out_unit_raster.save(r"C:\Users\57320\Dropbox\Mi PC (LAPTOP-5K2HI422)\Documents\ArcGIS for Climate Change\temperature\MOHC-HadGEM2-ES_tas.crf")

In [79]:
# precipitation
prcp_CanESM2 = Raster(r"C:\Users\57320\Dropbox\Mi PC (LAPTOP-5K2HI422)\Documents\ArcGIS for Climate Change\precipitation\CCCma-CanESM2_prcp.crf", True)
prcp_HadGEM2 = Raster(r"C:\Users\57320\Dropbox\Mi PC (LAPTOP-5K2HI422)\Documents\ArcGIS for Climate Change\precipitation\MOHC-HadGEM2-ES_prcp.crf", True)
prcp_MIROC5 = Raster(r"C:\Users\57320\Dropbox\Mi PC (LAPTOP-5K2HI422)\Documents\ArcGIS for Climate Change\precipitation\MIROC-MIROC5_prcp.crf", True)
# temperature
tas_CanESM2 = Raster(r"C:\Users\57320\Dropbox\Mi PC (LAPTOP-5K2HI422)\Documents\ArcGIS for Climate Change\temperature\CCCma-CanESM2_tas.crf", True)
tas_HadGEM2 = Raster(r"C:\Users\57320\Dropbox\Mi PC (LAPTOP-5K2HI422)\Documents\ArcGIS for Climate Change\temperature\MOHC-HadGEM2-ES_tas.crf", True)
tas_MIROC5 = Raster(r"C:\Users\57320\Dropbox\Mi PC (LAPTOP-5K2HI422)\Documents\ArcGIS for Climate Change\temperature\MIROC-MIROC5_tas.crf", True)

## Create File Geodatabase

In [84]:
data_path = "C:/Users/57320/Dropbox/Mi PC (LAPTOP-5K2HI422)/Documents/ArcGIS for Climate Change" 
gdb_name = "Data_precipitation.gdb"

# Run CreateFileGDB
arcpy.management.CreateFileGDB(data_path, gdb_name)

data_path = "C:/Users/57320/Dropbox/Mi PC (LAPTOP-5K2HI422)/Documents/ArcGIS for Climate Change" 
gdb_name = "Data_temperature.gdb"

# Run CreateFileGDB
arcpy.management.CreateFileGDB(data_path, gdb_name)

<Result 'C:/Users/57320/Dropbox/Mi PC (LAPTOP-5K2HI422)/Documents/ArcGIS for Climate Change/Data_temperature.gdb'>